packages/webapp/src/util/task.ts

```diff
@@ -0,0 +1,203 @@
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { SoilAmendmentPurpose } from '../store/api/types';
+import { getUnitOptionMap } from './convert-units/getUnitOptionMap';
+
+interface UnitOption {
+  label: string;
+  value: string;
+}
+
+interface PurposeRelationship {
+  purpose_id: number;
+  other_purpose?: string;
+}
+
+type DBSoilAmendmentTaskProduct = {
+  purpose_relationships: PurposeRelationship[];
+  other_purpose?: string;
+  weight?: number;
+  weight_unit?: string;
+  volume?: number;
+  volume_unit?: string;
+  percent_of_location_amended: number;
+  total_area_amended: number;
+  application_rate_weight?: number;
+  application_rate_weight_unit?: string;
+  application_rate_volume?: number;
+  application_rate_volume_unit?: string;
+  [key: string]: any;
+};
+
+type FormSoilAmendmentTaskProduct = {
+  purposes: number[];
+  other_purpose?: string;
+  weight?: number;
+  weight_unit?: UnitOption;
+  volume?: number;
+  volume_unit?: UnitOption;
+  percent_of_location_amended: number;
+  total_area_amended: number;
+  total_area_amended_unit?: UnitOption;
+  application_rate_weight?: number;
+  application_rate_weight_unit?: UnitOption;
+  application_rate_volume?: number;
+  application_rate_volume_unit?: UnitOption;
+  is_weight: boolean;
+  [key: string]: any;
+};
+
+type DBSoilAmendmentTask = {
+  soil_amendment_task_products: DBSoilAmendmentTaskProduct[];
+  [key: string]: any;
+};
+
+type FormSoilAmendmentTask = {
+  soil_amendment_task_products: FormSoilAmendmentTaskProduct[];
+  [key: string]: any;
+};
+
+export const formatSoilAmendmentTaskToFormStructure = (
+  task: DBSoilAmendmentTask,
+): FormSoilAmendmentTask => {
+  const taskClone = structuredClone(task);
+
+  const formattedTaskProducts = task.soil_amendment_task_products.map(
+    (dbTaskProduct: DBSoilAmendmentTaskProduct): FormSoilAmendmentTaskProduct => {
+      const { purpose_relationships, ...rest } = dbTaskProduct;
+      const isWeight = !!(rest.weight || rest.weight === 0);
+
+      const formattedTaskProduct = {
+        ...rest,
+        purposes: [],
+        is_weight: isWeight,
+      } as FormSoilAmendmentTaskProduct;
+
+      dbTaskProduct.purpose_relationships.forEach(({ purpose_id, other_purpose }) => {
+        if (other_purpose) {
+          formattedTaskProduct.other_purpose = other_purpose;
+        }
+        formattedTaskProduct.purposes.push(purpose_id);
+      });
+
+      const unitOptions: { [key: string]: UnitOption } = getUnitOptionMap();
+
+      return {
+        ...formattedTaskProduct,
+        weight_unit: isWeight && rest.weight_unit ? unitOptions[rest.weight_unit] : undefined,
```

Comment:
@SayakaOno for the systems-switching bug I was thinking of passing `system` here into this function and doing the 

```
soilAmounts[system].units.includes(weight_unit)
```
Do you have a preference about whether it's done here or in the component?

Reply:
System switching has been handled in the useUnit, so we shouldn't need to do it anywhere...
What was the issue?

Reply:
Ah, you know what the cause was? It's a bit of a fun/subtle one 🙂 

Unit does handle the system switching, but it needs to receive a string here, not a `UnitOption`. If the line above is changed to

`weight_unit: isWeight ? rest.weight_unit : undefined,`

then the system conversion works as expected without anything further. I'll update on my branch.

Reply:
Ah sorry 😓 Thank you!!!

Reply:
I don't know why why this should be. It looks in the `useUnit` code like both `UnitOption` and string should behave the same. But the system switching wants a string, and, to make it worse, `setValue()` only works with a `UnitOption`! 😂 What an unintuitive situation!